### DISCUSSION TITLE
안녕하세요 질문드립니다.

### DISCUSSION THREAD

**Discussion by uigl777** [nsoojin/MiniSuperApp-fastcampus#16] [아무 질문이나 환영합니다] [answered]
Swift UI를 사용하면 RIBs를 사용할 필요 없다고 하는데,
같이 사용하기에는 힘든가요?

**Comment by nsoojin** [comment, marked as answer]
안녕하세요.
RIBs에서는 interactor에서 UI를 업데이트하고 싶을때 presenter(UI)를 직접 호출하는데 반해 SwiftUI는 데이터에 바인딩이 되어 있으니 데이터를 변경하면 UI가 알아서 갱신되는 형태입니다. 화면 전환도 마찬가지고요. 이런 이유로 RIBs는 imperative UI에 적합한데 SwiftUI는 declarative UI이기에 궁합이 썩 좋지는 않습니다.

**Reply by uigl777** [reply]
RIBs를 공부하면 앞으로 많은 도움이 되겠죠?
곧 Swift UI가 대세가 될텐데, 불필요한 공부 시간이 될까 조바심이 나서 질문드려봅니다..

**Reply by nsoojin** [reply]
안녕하세요. 좋은 질문입니다. 그 마음도 정말 이해가 가고요.ㅎㅎ
저는 iOS개발 처음 시작하고 다음해에 스위프트가 나와서 옵젝씨 배운거 쓸모 없어진거 아닌가, 빨리 옵젝씨 버리고 스위프트 배워야하는거 아닌가 하는 조바심에 사로 잡혔던 추억이 기억나네요.

결론부터 말씀드리면 RIBs를 지금 하셔도 많은 도움이 될겁니다. SwiftUI냐 UIKit이냐는 별로 중요하지 않습니다. 특성이 다른 UI 프레임워크일 뿐이고요, 좋은 코드와 좋은 아키텍처를 만드는 원리 원칙은 프레임워크와 무관합니다. 다시 말해 SwiftUI를 쓰든 UIKit을 쓰든, imperative 또는 declarative를 쓰든지 코딩을 잘하는 핵심은 비슷하다는 겁니다. 제 아키텍처 강의에서 제일 강조하고자 했던 것 딱 하나가 composition입니다. 이건 iOS 개발을 할때 뿐 아니라 다른 개발을 할때도 중요합니다. 제가 해본적도 없던 플러터(SwiftUI처럼 declarative UI)를 맡고, 심지어 리딩해야했을 때 composition의 원리와 장점을 알았던 것이 정말 큰 도움이 됐습니다.

그리고 아직까지 SwiftUI와 잘 맞다고 알려지고 검증된 좋은 아키텍처가 없습니다. iOS 개발자들의 경험이 더 쌓이고 사례들이 나와야해서 시간이 더 필요할 것으로 보입니다. 제가 느끼기에 현재 상황은 SwiftUI에 열광하는 시기를 지나, 부작용을 토로하는 글들이 속속 보이기 시작합니다. MVVM이나, 애플이 알려주는 간단한 방식으로는 복잡한 상태관리가 어렵다는 것이죠. 그냥 SwiftUI를 써보고 싶어서 쓰는건 상관없지만, 아키텍처와 기본기를 다지는 목적으로는 최선의 선택은 아니라고 보입니다.

RIBs를 통해 composition과 testability를 습득하시면 나중에 무슨 UI 프레임워크를 쓰든, 심지어 나중에 플러터를 하거나 다른 플랫폼을 개발하더라도 자산이 될거라 생각합니다. 저의 경험을 증거로 말씀드릴수 있습니다.

**Reply by uigl777** [reply]
안녕하세요. 답변 정말 감사합니다^^
강의 보구 열심히 공부하겠습니다.
